src/qpx_mmc_css.cpp

```diff
@@ -0,0 +1,1909 @@
+
+  /*
+   * This encryption engine implements one of 32 variations
+   * one the same theme depending upon the choice in the
+   * variant parameter (0 - 31).
+   *
+   * The algorithm itself manipulates a 40 bit input into
+   * a 40 bit output.
+   * The parameter 'input' is 80 bits.  It consists of
+   * the 40 bit input value that is to be encrypted followed
+   * by a 40 bit seed value for the pseudo random number
+   * generators.
+   */
+
+  /* Feed the secret into the input values such that
+   * we alter the seed to the LFSR's used above,  then
+   * generate the bits to play with.
+   */
+  for (i = 5; --i >= 0;) {
+    p_tmp1[i] = p_scratch[5 + i] ^ p_secret[i] ^ p_crypt_tab2[i];
+  }
+
+  /*
+   * We use two LFSR's (seeded from some of the input data bytes) to
+   * generate two streams of pseudo-random bits.  These two bit streams
+   * are then combined by simply adding with carry to generate a final
+   * sequence of pseudo-random bits which is stored in the buffer that
+   * 'output' points to the end of - len is the size of this buffer.
+   *
+   * The first LFSR is of degree 25,  and has a polynomial of:
+   * x^13 + x^5 + x^4 + x^1 + 1
+   *
+   * The second LSFR is of degree 17,  and has a (primitive) polynomial of:
+   * x^15 + x^1 + 1
+   *
+   * I don't know if these polynomials are primitive modulo 2,  and thus
+   * represent maximal-period LFSR's.
+   *
+   *
+   * Note that we take the output of each LFSR from the new shifted in
+   * bit,  not the old shifted out bit.  Thus for ease of use the LFSR's
+   * are implemented in bit reversed order.
+   *
+   */
+
+  /* In order to ensure that the LFSR works we need to ensure that the
+   * initial values are non-zero.  Thus when we initialise them from
+   * the seed,  we ensure that a bit is set.
+   */
+  i_lfsr0 = (p_tmp1[0] << 17) | (p_tmp1[1] << 9) | ((p_tmp1[2] & ~7) << 1) | 8 |
+            (p_tmp1[2] & 7);
+  i_lfsr1 = (p_tmp1[3] << 9) | 0x100 | p_tmp1[4];
+
+  i_index = sizeof(p_bits);
+  i_carry = 0;
+
+  do {
+    for (i_bit = 0, i_val = 0; i_bit < 8; ++i_bit) {
+
+      i_lfsr0_o = ((i_lfsr0 >> 24) ^ (i_lfsr0 >> 21) ^ (i_lfsr0 >> 20) ^
+                   (i_lfsr0 >> 12)) &
+                  1;
+      i_lfsr0 = (i_lfsr0 << 1) | i_lfsr0_o;
+
+      i_lfsr1_o = ((i_lfsr1 >> 16) ^ (i_lfsr1 >> 2)) & 1;
+      i_lfsr1 = (i_lfsr1 << 1) | i_lfsr1_o;
+
+      i_combined = !i_lfsr1_o + i_carry + !i_lfsr0_o;
+      /* taking bit 1 */
+      i_carry = (i_combined >> 1) & 1;
+      i_val |= (i_combined & 1) << i_bit;
+    }
+
+    p_bits[--i_index] = i_val;
+  } while (i_index > 0);
+
+  /* This term is used throughout the following to
+   * select one of 32 different variations on the
+   * algorithm.
+   */
+  i_cse = p_variants[i_css_variant] ^ p_crypt_tab2[i_css_variant];
+
+  /* Now the actual blocks doing the encryption.  Each
+   * of these works on 40 bits at a time and are quite
+   * similar.
+   */
+  i_index = 0;
```

Comment:
Codacy found an issue: [Variable 'i_index' is assigned a value that is never used.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)